DOCS-3192: Add discovery service page (#4043)
viamrobotics/docs: docs/operate/reference/services/discovery/_index.md

---
title: "Discovery Service"
linkTitle: "Discovery"

> description: "The discovery service is used to discover resources on a machine."

Reviewer: passive voice

type: docs
weight: 50
no_list: true
#icon: true
#images: ["/services/icons/discovery.svg"]
tags: ["discovery", "services"]
no_service: true
date: "2025-03-03"
# updated: ""  # When the content was last entirely checked
# SMEs: John N.
modulescript: true
---

> The discovery service is used to discover resources on a machine.

Reviewer: passive voice

Reviewer: Does the discovery service discover resources as in components OR does it discover attached hardware (and the paths under which the hardware is available) or both? Clarifying this further is I think important.

If you are [creating a modular resource](/operate/get-started/other-hardware/) that depends on other resources that are discoverable in a systematic way, you can create a discovery service as part of your module to discover those resources.

> ## Example usage

Reviewer: This is all a bit abstract to me right now. What does it discover/return and how do I use it? Do I need to view that in the logs or does the UI just add a drop down? Since this is already covering example usage, can you walk through a specific vision service that does this? Then you can explain what the discovery service returns and how you can use it.

Author: Okay I added some more detail here. On Monday I'll check with John whether there's a better sort of example--right now there's rtsp and webcam, but not sure we want to go that specific?

Imagine you are creating a vision service module that depends on a camera.
You might want to include both a vision service and a discovery service in your module.
You could write a discovery service that reports all camera paths recognized by your computer or SBC.

> Users of your module could then:

Reviewer: Avoid conditional in favour of explaining the value this would add up front. 
Vale should be underlining conditionals and passive voice for you. Please check it's working.

1. Configure the discovery service in their machine's configuration.
1. Use it to discover any cameras that are connected to the machine.

> 1. Use one of the discovered cameras in their vision service configuration.

Reviewer 2: [q] so visually it looks like this works, but I notice in some examples we step through numbers ourselves(ex 1. 2. 3.) is it preferred to leave it like this?

Author: These render correctly and I like doing it this way since when you add or delete a step you don’t have to change them all. Thanks for checking!

---
title: "Discovery Service"
linkTitle: "Discovery"
description: "Use a discovery service to discover available resources on a machine."
type: docs
weight: 50
no_list: true
#icon: true
#images: ["/services/icons/discovery.svg"]
tags: ["discovery", "services"]
no_service: true
date: "2025-03-03"
# updated: ""  # When the content was last entirely checked
# SMEs: John N.
modulescript: true
---

A discovery service allows you to return a list of physical hardware available on a machine, and suggest configurations for those components to integrate the hardware into the machine.

> If you are [creating a modular resource](/operate/get-started/other-hardware/) that depends on other resources that are discoverable in a systematic way, you can create a discovery service as part of your module to discover those resources.

Reviewer: I think also here we want to be more specific than resources. If it's hardware then I'd say hardware resources and if it's viam resources then I think add the glossary entry so there's no room for misinterpretation. Still not sure whether it's meant just for hardware or for hardware or configured components/services

Author: I mean, it is just an API. So someone could use it differently. I believe the idea for it was to surface hardware, but you could implement it differently--just have to return a list of resource configs.

> If you are [creating a modular resource](/operate/get-started/other-hardware/) that depends on other resources that are discoverable in a systematic way, you can create a discovery service as part of your module to discover those resources.

Author: ```suggestion
If you are [creating a modular resource](/operate/get-started/other-hardware/) that depends on other {{< glossary_tooltip term_id="resource" text="resources" >}} that are discoverable in a systematic way, you can create a discovery service as part of your module to discover those resources.
```

## Example usage

> Imagine you are creating a vision service module that depends on a camera, and you want to make it easier for users to configure the camera.
> You can include both a vision service and a discovery service in your module.
> You'd write a discovery service that reports all camera paths recognized by your computer or SBC.
> Users of your module could then:

Reviewer: Remove conditional & shorten
```suggestion
Imagine you are creating a vision service module that depends on a camera.
To make it easier for users to configure the camera, you include a discovery service in your module.
You implement the discovery service to report all the camera paths your computer or SBC finds.
Users of your module can then:
```

1. Configure both the vision service and the discovery service in their machine's configuration.
1. Click the **Test** panel in the discovery service configuration to see all cameras recognized by the machine, presented as a list of configuration snippets.

> 1. Copy the copy-pasteable configuration snippet for the camera they want to use, with the camera path already filled in.

Reviewer: ```suggestion
1. Create a camera with the copy-pasteable configuration snippet for the camera they want to use, with the camera path already filled in.
```

> 1. Use one of the discovered cameras in their vision service configuration.

Reviewer: ```suggestion
1. Use the configured camera in their vision service configuration.
```
